### PR TITLE
Assemble changelog only in PRs comments

### PR DESCRIPTION
**Pull request by SevaZhukov**
After further consideration, it has been determined that it is not possible to both assemble and push the changelog to protected branches. Using PRs to update the changelog would simply bring us back to the original issue. As a result, the script has been refactored to only add the assembled changelog to PRs comments.

### CODE REVIEW

**Comment by dzinad** [comment]
> After further consideration, it has been determined that it is not possible to both assemble and push the changelog to protected branches.

Why? Is this because of some restrictions on the main branch?

**Comment by dzinad** [comment]
Can we print the changelog automatically in the PR? Like we do with codecov? Because I seriously doubt that someone will manually run the script.

**Comment by VysotskiVadim** [comment]
BTW, what is the flow now. I add a single changelog entry and merge pr without updating CHANGELOG.md. During release we assemble changelog changelog from entries and remove them, right?

**Comment by SevaZhukov** [comment]
> release train will assemble the changelog automatically and open and merge the PR, right? Or release duty will have to do it manually?

the release train will do it automatically

**Comment by SevaZhukov** [comment]
> BTW, what is the flow now. I add a single changelog entry and merge pr without updating CHANGELOG.md. During release we assemble changelog changelog from entries and remove them. is my understanding correct?

yes, I mentioned it [here](https://github.com/mapbox/mapbox-navigation-android/blob/szh-changelog-strategy-fixes/changelog/README.md)

File: changelog/README.md [hunk 1]
```diff
@@ -47,11 +45,11 @@ Every push to the main or release branch Assemble changelog GitHub action will b
 Some other changes
 ```
 
-* write the changelog to the `changelog/unreleased/CHANGELOG.md` file
+To view the changelog for the android auto project, run the command `python3 scripts/changelog/assemble_changelog.py --auto`.
 
 Every release the release train app will:
 
-* get changelog from `changelog/unreleased/CHANGELOG.md` file
+* assemble the changelog by the script `python3 scripts/changelog/assemble_changelog.py`
```
Comment (VysotskiVadim) [inline changelog/README.md:52]: If you add `#!/usr/bin/env python3` to the top of the file, you can run script typing less 🙂 

```suggestion
* assemble the changelog by the script `./scripts/changelog/assemble_changelog.py`
```

**Review by VysotskiVadim — APPROVED** [review]

**Review by dzinad — APPROVED** [review]

**Comment by abhishek1508** [comment]
Do we need to assemble the changelog with every PR? If we are creating a separate `*.md` file with every PR, that hosts the change introduced, can we just not assemble all the changes by going through each `*.md` and adding them directly to `root/CHANGELOG.md` at the time of release?

I am not in favor of adding the assembled changelog to PR comment as I am not sure what value am I getting out of it. This also pollutes the PR in case there are several discussions happening which by the way is most often the case based on code review. The changes in `README.md` mentions:

> The comment will be updated with every change.

Does this mean the assembled changelog will be posted as PR comment every time a new commit is pushed or forced rebase is made?

**Comment by SevaZhukov** [comment]
> We could be a bit more strict with this repo permissions. Grant "navigation team" members the "Maintain" role, a couple leads Admin role and automation accounts Custom role with the Bypass branch protections permission.

We can't give the bot permission **Bypass branch protections** anyway. Because the script uses the github-actions token, it's not a user.

**Comment by SevaZhukov** [comment]
> Does this mean the assembled changelog will be posted as PR comment every time a new commit is pushed or forced rebase is made?

Yes, but it will create a comment if it doesn't exist or update if it exists.

**Comment by SevaZhukov** [comment]
I propose a compromise - a collapsible changelog comment. It doesn't take up a lot of space and it exists

**Comment by abhishek1508** [comment]
> I propose a compromise - a collapsible changelog comment. It doesn't take up a lot of space and it exists

This is a better solution in my opinion. Thanks for giving it a deeper thought. 

However, I still don't understand the need for assembling the changelog in every PR and putting it as a PR comment. Can someone explain the benefit of it?

cc @mapbox/navigation-android

**Comment by tomaszrybakiewicz** [comment]
> > We could be a bit more strict with this repo permissions. Grant "navigation team" members the "Maintain" role, a couple leads Admin role and automation accounts Custom role with the Bypass branch protections permission.
> 
> We can't give the bot permission **Bypass branch protections** anyway. Because the script uses the github-actions token, it's not a user.

@SevaZhukov Can't we request a separate user, assign a custom role with **Bypass branch protections** and use its personal token for the workflow as suggested here? https://docs.github.com/en/actions/security-guides/automatic-token-authentication#granting-additional-permissions 
Of course, we would still need to downgrade everyone to the "Maintainer" role and disable "Do not allow bypassing the above settings" branch protection to make it work.

**Comment by SevaZhukov** [comment]
> Can't we request a separate user, assign a custom role with Bypass branch protections and use its personal token for the workflow as suggested here?

Sorry but no :( Because as I know we can't use and store permanent GitHub tokens. We can only get it by 1. mbx-ci (it doesn't work for this case because it's a public repo) 2. use default github-token for github-actions

**Comment by VysotskiVadim** [comment]
> However, I still don't understand the need for assembling the changelog in every PR and putting it as a PR comment. Can someone explain the benefit of it?

I think it's nice to see how new changelog entries affect the way whole changelog looks like. It may be useful if you write not a typical message, for instance with code examples or warnings about known issues.

**Pull request merged**